storage/country_tree.hpp

```diff
@@ -6,205 +6,234 @@
 #include "std/unique_ptr.hpp"
 #include "std/vector.hpp"
 
-/// This class is developed for using in Storage. It's a implementation of a tree.
+template <class K>
+struct NodeKeyHasher
+{
+  size_t operator()(K const & k) const { return m_hash(k); }
+private:
+  hash<K> m_hash;
+};
+
+/// This class is developed for using in Storage. It's a implementation of a tree with ability
+/// of access to its nodes in constant time with the help of hash table.
 /// It should be filled with AddAtDepth method.
 /// This class is used in Storage and filled based on countries.txt (countries_migrate.txt).
 /// While filling CountryTree nodes in countries.txt should be visited in DFS order.
-template <class T>
+template <class I, class T>
```

Comment:
I propose you to use more readable names, like TKey and TValue, or something like that.